nixos/grafana: documentation/warning improvements after #191768

Ma27 opened the pull request:
###### Description of changes

Further details are outlined in the commit messages. This basically
* makes sure that all sub-options appear in the manual
* the release-notes contain more details on the upgrade-path.
* ensure that warnings about potential secrets being leaked to the store are correct.

Follows up on #191768.

Closes #198646

cc @dhess @sternenseemann @KFearsoff @SuperSandro2000 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Ma27 committed:
nixos/grafana: `provision.{datasources,dashboards}` can't be a list anymore

The hack with `either` had the side-effect that the sub-options of the
submodule didn't appear in the manual. I decided to remove this because
the "migration" isn't that hard, you just need to fix some module
declarations.

However, `mkRenamedOptionModule` wouldn't work here because it'd create
a "virtual" option for the deprecated path (i.e.
`services.grafana.provision.{datasources,dashboards}`), but that's the
already a new option, i.e. the submodule for the new stuff.

To make sure that you still get errors, I implemented a small hack using
`coercedTo` which throws an error if a list is specified (as it would be
done on 22.05) which explains what to do instead to make the migration
easier.

Also, I linkified the options in the manual now to make it easier to
navigate between those.

Ma27 committed:
nixos/doc: improve release-notes for services.grafana

KFearsoff commented on nixos/doc/manual/from_md/release-notes/rl-2211.section.xml:
```diff
@@ -974,22 +974,129 @@
+                <para>
+                  Instead of declaring datasources and dashboards in
+                  pure Nix, it’s also possible to specify configuration
+                  directories with YAML instead using
```
Actually, you don't *have to* specify directories. I'm not sure if you even **can** specify directories in this case, since directories will be named `datasources.yaml` and I'm not sure if Grafana checks if `*.yaml` is actually a file or not.
It is probably a good idea to test what behaviours we support and what behaviours we don't support, and change either this changelog entry or the description of `path` options accordingly.

Ma27 replied:
Good catch, we only support single files here.

KFearsoff commented on nixos/tests/grafana/provision/default.nix:
```diff
@@ -192,14 +168,6 @@ in {
 
 
-
-    with subtest(f"Successful notifiers provision with {nodeOld[0]}"):
```
We haven't actually deprecated notifiers provisioning completely, we just put up a warning about it. Please don't remove the test. We can still test the behaviour if we move the notifiers provisoning into `nodeNix`.
Also, since we removed the `nodeOld`, we can now write all of the tests inside of the `for nodeInfo in [nodeNix, nodeYaml]` loop. That would look quite a bit cleaner.

KFearsoff reviewed (changes requested):
Haven't tested it. The docs look good to me. Some things are still a little up in the air, I have added comments on what I have doubts with.

Ma27 commented on nixos/doc/manual/from_md/release-notes/rl-2211.section.xml:
```diff
@@ -974,22 +974,129 @@
+            </para>
+            <para>
+              Alternatively you can also set all your values from
+              <literal>extraOptions</literal> to
```
Ohh you're absolutely right, I forgot about that!

(The documentation can stay as-is though I think).

KFearsoff replied:
Yup, no need to change documentation here. It's great.

Ma27 replied:
Fixed. Will look at the other comments later :)

Ma27 committed:
nixos/grafana: mark `services.grafana.extraOptions` as removed

KFearsoff commented on nixos/modules/services/monitoring/grafana.nix:
```diff
@@ -60,6 +48,26 @@ let
+  # FIXME(@Ma27) remove before 23.05. This is just a helper-type
+  # because `mkRenamedOptionModule` doesn't work if `foo.bar` is renamed
+  # to `foo.bar.baz`.
+  submodule' = module: types.coercedTo
```
Doesn't that mean that we can remove the warning for `datasources` and `dashboards`?
(I actually noticed that the warning for dashboards triggers on datasources instead. Sigh, could've sworn I've fixed it)

Ma27 committed:
nixos/grafana: re-add legacy notifiers test, mention notifiers in release notes

Ma27 committed:
nixos/grafana: get rid of unnecessary `flatten` for warnings

Ma27 committed:
nixos/grafana: we only support single YAML files for provisioning

KFearsoff commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -304,9 +304,65 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
+      this feature will is deprecated by Grafana and will probably removed in Grafana 10.
+      It's recommended to use `services.grafana.provision.alerting.contactPoints` instead.
+
+  - It's possible to configure Grafana's new alerting via `services.grafana.provision.alerting`.
```
We have a duplicate doc entry two lines below.

KFearsoff reviewed (approved):
Looks good to me, just change the duplicating entry and I think we're ready to go.

KFearsoff commented:
Ah, while we're at it, do you mind also adding the changes from #198646?
I'd like it to be explicit that we actually CAN use file provider when provisioning. It's just not documented upstream for some reason.

MinerSebas commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -304,9 +304,63 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
+      provisioning files from non-NixOS Grafana instances that you also want to
+      deploy to NixOS.
+
+    - [](#opt-services.grafana.provision.notifiers) is not affected by this change because
+      this feature will is deprecated by Grafana and will probably removed in Grafana 10.
+      It's recommended to use `services.grafana.provision.alerting.contactPoints` instead.
```
```suggestion
    - [](#opt-services.grafana.provision.notifiers) is not affected by this change because
      this feature is deprecated by Grafana and will probably removed in Grafana 10.
      It's recommended to use `services.grafana.provision.alerting.contactPoints` instead.
```

Ma27 committed:
nixos/tests/grafana: demonstrate how to use the file provider

Yes, it's still a store path, but it's only for demonstration purposes.
Also we now have fewer warnings in the provision test.

Ma27 committed:
nixos/doc: wording fix

Ma27 committed:
nixos/grafana: fix secret-related warnings

Closes #198646

* The options `password`/`basicAuthPassword` were removed for
  datasources in Grafana 9. The only option to declare them now is to use
  `secureJsonData`.
* Fix description for contactPoints provisioning: when using file/env
  providers, nothing will be leaked into the store.
* Fix regex in file-provider usage check: it's also possible to either
  use `$__env{FOO}` or `$FOO` to fetch secrets from the environment.
* Fix warning for datasources: `password`/`basicAuthPassword` was
  removed, also check for each setting in `secureJsonData` if
  env/file-provider was used (then no warning is needed!).

Ma27 committed:
nixos/grafana: fix w/o datasources or dashboard provisioning

Ma27 commented:
Took the liberty to mark this as blocker: the doc fixes are actually relevant for a smooth upgrade path of Grafana.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1383

SuperSandro2000 reviewed (commented):
I read through the changelog and didn't notice anything but I am not sure if it is complete. Not sure if the module is now working better and all the options are displayed on search.nixos.org.

Ma27 commented:
They should as they also appear in the manual.

KFearsoff commented on nixos/modules/services/monitoring/grafana.nix:
```diff
@@ -1159,52 +1166,50 @@ in {
-      (optional (
-        builtins.isList cfg.provision.datasources && cfg.provision.dashboards != []
+        Grafana passwords will be stored as plaintext in the Nix store!
+        Use file/env provider or an env-var instead.
```
I think you were trying to say `Use file provider or an env provider with an env-var instead`?

Ma27 replied:
Both exist, but to avoid confusion, I did `s|file/env|file|g` on the warnings. Good point.

KFearsoff reviewed (commented):
I've been thinking about it, and while it is kinda out of scope of this PR, but don't you think `*.path` options are completely useless? No one is gonna build one huge YAML for everything. People usually have a folder of alerting YAMLs.
Do you think we can remove `alerting.*.path` setttings, add `alerting.path` setting that would point to a directory and symlink it to `$out/provisioning/alerting` like [upstream does](https://grafana.com/docs/grafana/latest/administration/provisioning/#alerting)? Then we would also be able to convert `alerting.*.settings` to `alerting.*`, which should be a nicer syntax.

Ma27 committed:
nixos/grafana: allow using both directories or single YAML files for non-Nix provisioning

Ma27 committed:
nixos/grafana: make warning more clear

Ma27 commented:
Good point, the paths that originally could only be YAML files can now point to directories as well.
I decided to not alter the option-tree even more because people have already started their migrations apparently and we should come to an end here.

@NixOS/release-engineers @NixOS/nixos-release-managers this should end up in 22.11 as it fixes a few remaining issues of the previous Grafana refactoring. Additional reviews would be fairly useful before merging that.

KFearsoff reviewed (approved):
Looks good to me, don't have any more inputs. Would be very nice to have some more reviews, cos I don't quite imagine releasing 22.11 without this.

lheckemann commented on nixos/modules/services/monitoring/grafana.nix:
```diff
@@ -13,57 +13,96 @@ let
   _filter = x: filterAttrs (k: v: k != "_module") x;
 
+  # FIXME(@Ma27) remove before 23.05. This is just a helper-type
+  # because `mkRenamedOptionMthere's there's odule` doesn't work if `foo.bar` is renamed
```
Looks like a weird paste happened here

lheckemann commented on nixos/modules/services/monitoring/grafana.nix:
```diff
@@ -330,19 +351,7 @@ in {
-                  ln -sf ''${muteTimingsFile} $out/alerting/muteTimings.yaml
-                  \'\'
-              '';
+              defaultText = "directory with all provisioning files linked together";
```
```suggestion
              defaultText = "directory with links to files generated from config.grafana.provision";
```
(ideally with the option linking syntax, don't know what that is nowadays)

lheckemann commented on nixos/tests/grafana/provision/default.nix:
```diff
@@ -157,6 +163,22 @@ let
+    };
   };
 
   nodes = builtins.mapAttrs (_: val: mkMerge [ val baseGrafanaConf ]) extraNodeConfs;
```
(not a blocker, fine to fix this later IMHO) wouldn't it make more sense to have 3 separate tests for this rather than a single, more resource-intensive test which goes through all 3 variants serially?

lheckemann reviewed (approved):
Looks sensible overall, haven't been able to test it yet though.

Ma27 committed:
nixos/grafana: review fixes

Ma27 committed:
nixos/doc: also note that external YAML files for grafana will end up in the store

Ma27 commented:
So it just occurred to me that it should most likely be possible to allow both YAML and Nix configurations simultaneously (when keeping the current option-structure inside `provisioning.alerting`). However, this should be implemented in a follow-up. This change ensures a smooth migration to the new module and thus should be part of the new release branch tomorrow, so I'll merge as soon as CI is finished.
